.gitlab-ci.yml

```diff
@@ -1,232 +1,302 @@
+  script:
+    - scripts/gitlab/test.sh nightly
+
+#test:rustfmt:
+#  <<:                             *optional_test
+#  script:
+#    - scripts/gitlab/rustfmt.sh
+
+#test:clippy:
+#  <<:                             *optional_test
+#  script:
+#    - scripts/gitlab/clippy.sh
+
+test:coverage:
+  stage:                           test
   only:
-    - beta
-    - tags
-    - stable
-    - triggers
+    - master
   script:
-    - scripts/gitlab-build.sh aarch64-unknown-linux-gnu aarch64-unknown-linux-gnu arm64 aarch64-linux-gnu-gcc aarch64-linux-gnu-g++ linux
+    - scripts/gitlab/coverage.sh
   tags:
-    - rust-arm
-  artifacts:
-    paths:
-    - parity.zip
-    name: "aarch64-unknown-linux-gnu_parity"
-linux-snap:
-  stage: build
-  image: parity/snapcraft:gitlab-ci
-  only:
-    - stable
-    - beta
-    - tags
-    - triggers
+    - shell
+  allow_failure:                   true
+
+
+#### stage:                        build
+
+build:linux:ubuntu:amd64:          &build
+  stage:                           build
+  only:                            *releaseable_branches
+  variables:
+    CARGO_TARGET:                  x86_64-unknown-linux-gnu
   script:
-    - scripts/gitlab-build.sh x86_64-unknown-snap-gnu x86_64-unknown-linux-gnu amd64 gcc g++ snap
+    - scripts/gitlab/build-unix.sh
+  <<:                              *collect_artifacts
   tags:
     - rust-stable
-  artifacts:
-    paths:
-    - parity.zip
-    name: "stable-x86_64-unknown-snap-gnu_parity"
-darwin:
-  stage: build
-  only:
-    - beta
-    - tags
-    - stable
-    - triggers
-  script:
-    - scripts/gitlab-build.sh x86_64-apple-darwin x86_64-apple-darwin macos gcc g++ macos
+  allow_failure: true
+
+build:linux:ubuntu:i386:
+  <<:                              *build
+  image:                           parity/rust-i686:gitlab-ci
+  variables:
+    CARGO_TARGET:                  i686-unknown-linux-gnu
+
+build:linux:ubuntu:arm64:
+  <<:                              *build
+  image:                           parity/rust-arm64:gitlab-ci
+  variables:
+    CARGO_TARGET:                  aarch64-unknown-linux-gnu
+
+build:linux:ubuntu:armhf:
+  <<:                              *build
+  image:                           parity/rust-armv7:gitlab-ci
+  variables:
+    CARGO_TARGET:                  armv7-unknown-linux-gnueabihf
+
+build:linux:android:armhf:
+  <<:                              *build
+  image:                           parity/rust-android:gitlab-ci
+  variables:
+    CARGO_TARGET:                  armv7-linux-androideabi
+
+build:macos:x86_64:
+  <<:                              *build
+  variables:
+    CARGO_TARGET:                  x86_64-apple-darwin
+    CC:                            gcc
+    CXX:                           g++
   tags:
     - osx
-  artifacts:
-    paths:
-    - parity.zip
-    name: "x86_64-apple-darwin_parity"
-windows:
+  <<:                              *collect_artifacts
+
+build-windows-x86_64:
```

Comment:
also `build:windows:msvc:x86_64`?

Reply:
Unfortunately, windows builds can't have `:` characters in their names, because windows paths can't have said character.

Maybe let's come up with a better (and universal) separator?

Reply:
Ok let's go back to `-` then